aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -240,6 +242,23 @@ private void newStream() throws IOException {
 
   @Override
   public void close() throws IOException {
+
+    /**
+     * As calls to close() are used to mark a stream as completed and add the file
+     * to the completed data files in BaseTaskWriter we cannot just return if this
+     * was already closed with failure to upload.
+     *
+     * As close 'completes' the upload and staging files are cleanup, any failure to
+     * upload should be persisted and thrown back to the caller to make sure
+     * the underlying file represented by this stream is not used in the completedDataFiles
+     * of a manifest.
+     * see: https://github.com/apache/iceberg/issues/5310
+      */
+    if(closeFailureState != null){
+      throw new IllegalStateException("close called on a closed stream which failed to close completely earlier at: "+closeFailureState.getFailureAt()
```

Comment:
Few style errors on this line, Checkstyle will mark them as errors though. (Missing space between + and string, "comma after a newline"). Instructions on running our current checkstyle rules are in the dev docs